packages/kilt-did/README.md

```diff
@@ -3,6 +3,339 @@
+
+By design, DID signatures and Substrate signatures are decoupled, meaning that the encoded and signed DID creation operation can then be signed and submitted by a different KILT account than the DID subject. We think this opens the path for a wider range of use cases in which, for instance, a service provider might be willing to offer a DID-as-a-Service option for its customers.
+
+An example of a full DID is the following:
+
+```
+did:kilt:4rp4rcDHP71YrBNvDhcH5iRoM3YzVoQVnCZvQPwPom9bjo2e
+```
+
+Here, there is no `light:` component, which indicates that the DID is a full DID and that the keys associated with it must not be derived from the DID identifier but must be retrieved from the KILT blockchain.
+
+Beyond an authentication key, and encryption key, and external services, a full DID also supports an **attestation key**, which must be used to write CTypes and attestations on the blockchain, and a **delegation key**, which must be used to write delegations on the blockchain.
+
+### Creating and anchoring a full DID
+
+The following is an example of how to create and write on blockchain a full DID that specifies only an authentication key.
+
+```typescript
+import Keyring from '@polkadot/keyring'
+
+import { BlockchainUtils } from '@kiltprotocol/chain-helpers'
+import { init as kiltInit } from '@kiltprotocol/core'
+import {
+  DefaultResolver,
+  DemoKeystore,
+  DidUtils,
+  FullDidDetails,
+  SigningAlgorithms
+} from '@kiltprotocol/did'
+import { KeyRelationship } from '@kiltprotocol/types'
+
+// Configure the resolution promise to wait for transactions to be finalized or simply included in a block depending on the environment.
+const resolveOn = process.env.NODE_ENV === 'production' ? BlockchainUtils.IS_FINALIZED : BlockchainUtils.IS_IN_BLOCK
+
+// Initialise connection to the public KILT test network.
+await kiltInit({ address: 'wss://kilt-peregrine-k8s.kilt.io' })
+
+// Generate the KILT account that will submit the DID creation tx to the KILT blockchain.
+// It must have enough funds to pay for the tx execution fees.
+const aliceKiltAccount = new Keyring({
+  type: 'ed25519',
+  // KILT has registered the ss58 prefix 38
+  ss58Format: 38,
+}).createFromUri('//Alice')
+
+// Instantiate the demo keystore.
+const keystore = new DemoKeystore()
+
+// Generate seed for the authentication key.
+const authenticationSeed = '0x123456789'
+
+// Ask the keystore to generate a new keypar to use for authentication.
+const authenticationKeyPublicDetails = await keystore.generateKeypair({
+  seed: authenticationSeed,
+  alg: SigningAlgorithms.Ed25519,
+})
+
+// Generate the DID-signed creation extrinsic.
+// The extrinsic is unsigned and contains the DID creation operation signed with the DID authentication key.
+const { extrinsic, did } = await DidUtils.writeDidFromPublicKeys(keystore, {
```

Comment:
C: It is weird to me, that for the off-chain DID, we use `new LightDidDetails`, but for the on-chain DID, we write it directly to the chain via the DidUtils and resolve it to get the `FullDidDetails`.

Reply:
Yes, for me as well. The thing is that the `FullDidDetails` constructor is really just a wrapper around chain data (e.g., the keys it takes in input contain info about the block at which the key was created). In the refactoring, we should probably make the constructor async (if that is allowed) and do both creation and writing on the blockchain at once. And keep this constructor for when we retrieve stuff from the blockchain.